### PR TITLE
Add Site Info Properties builder

### PR DESCRIPTION
**Pull request by smarchenko**
Use SiteInfoPropertyBuilder to create SiteInfo properties dictionary.
Use new extension methods to convert builder to SiteInfo, Site and SiteContext

### CODE REVIEW

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
+﻿using Sitecore.Data;
+using StringDictionary = Sitecore.Collections.StringDictionary;
+
+namespace Sitecore.NSubstituteUtils
+{
+    public class SiteInfoPropertiesBuilder
+    {
+        public SiteInfoPropertiesBuilder()
```
Comment (mitikov) [inline code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs]: What is the mainstream API usage per your vision?

To me - setting the site name, or/and database name.

Is it possible to introduce constructor with optional parameters (f.e. siteName), similar to FakeItem has?

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+        public SiteInfoPropertiesBuilder WithRequireLogin(bool requireLogin)
+        {
+            SiteInfoProperties[Constants.SiteProperties.RequireLogin] = requireLogin.ToString().ToLowerInvariant();
+            return this;
+        }
+
+        public StringDictionary ToSitecoreSiteInfoProperties()
+        {
+            return SiteInfoProperties;
```
Comment (mitikov) [inline code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs]: Could we [return a copy of the dictionary](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/guidelines-for-collections#collection-properties-and-return-values) instead?

The use-case is to:
1) Arrange one Site into 
2) Build it
3) Do a few more arrange steps
4) Build a new one 
**Actual**: step 2 and step 4 point on the same object
**Expected**: elements do not depend on each other

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,186 @@
+﻿using Sitecore.Data;
+using StringDictionary = Sitecore.Collections.StringDictionary;
+
+namespace Sitecore.NSubstituteUtils
+{
+    public class SiteInfoPropertiesBuilder
```
Comment (mitikov) [inline code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs:9]: It is very convenient to use implicit casts for FakeItem -> Item.
Could we have a similar operator for SiteInfoPropertiesBuilder?

**Review by mitikov — COMMENTED** [review]
I like the overall idea and implementation.

Please take a look at a few comments.

Thanks.

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs (hunk 2 above)
Comment (mitikov) [reply code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs]: I'm thinking to leave as is and implement IClonable to achieve the same outcome.
I would use: 
Item item = fakeItem
fakeItem.WithField(...)

And item gets that field set; I think that keeping the behavior is more important in this case.

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs (hunk 1 above)
Comment (smarchenko) [reply code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs]: Done. Constructor that accepts site name has been introduced

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs (hunk 2 above)
Comment (smarchenko) [reply code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs]: Done. Flag has been introduced. Returns original collection by default

File: code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs (hunk 3 above)
Comment (smarchenko) [reply code/Sitecore.NSubstitute/SiteInfoPropertiesBuilder.cs:9]: Done. Implicit casts have been added

**Review by mitikov — APPROVED** [review]

**Pull request merged**